pyteal/ast/router.py

```diff
@@ -0,0 +1,634 @@
+    - and a contract JSON object allowing for easily read and call methods in the contract
+
+    *DISCLAIMER*: ABI-Router is still taking shape and is subject to backwards incompatible changes.
+
+    * Based on feedback, the API and usage patterns are likely to change.
+    * Expect migration issues.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        bare_calls: BareCallActions = None,
+    ) -> None:
+        """
+        Args:
+            name: the name of the smart contract, used in the JSON object.
+            bare_calls: the bare app call registered for each on_completion.
+        """
+
+        self.name: str = name
+        self.approval_ast = ASTBuilder()
+        self.clear_state_ast = ASTBuilder()
+
+        self.method_sig_to_selector: dict[str, bytes] = dict()
+        self.method_selector_to_sig: dict[bytes, str] = dict()
+
+        if bare_calls and not bare_calls.is_empty():
+            bare_call_approval = bare_calls.approval_construction()
+            if bare_call_approval:
+                self.approval_ast.conditions_n_branches.append(
+                    CondNode(
+                        Txn.application_args.length() == Int(0),
+                        cast(Expr, bare_call_approval),
+                    )
+                )
+            bare_call_clear = bare_calls.clear_state_construction()
+            if bare_call_clear:
+                self.clear_state_ast.conditions_n_branches.append(
+                    CondNode(
+                        Txn.application_args.length() == Int(0),
+                        cast(Expr, bare_call_clear),
+                    )
+                )
+
+    def add_method_handler(
+        self,
+        method_call: ABIReturnSubroutine,
+        overriding_name: str = None,
+        method_config: MethodConfig = MethodConfig(),
+    ) -> None:
+        if not isinstance(method_call, ABIReturnSubroutine):
+            raise TealInputError(
+                "for adding method handler, must be ABIReturnSubroutine"
+            )
+        method_signature = method_call.method_signature(overriding_name)
+        if method_config.is_never():
+            raise TealInputError(
+                f"registered method {method_signature} is never executed"
+            )
+        method_selector = encoding.checksum(bytes(method_signature, "utf-8"))[:4]
+
+        if method_signature in self.method_sig_to_selector:
+            raise TealInputError(f"re-registering method {method_signature} detected")
+        if method_selector in self.method_selector_to_sig:
+            raise TealInputError(
+                f"re-registering method {method_signature} has hash collision "
+                f"with {self.method_selector_to_sig[method_selector]}"
+            )
+        self.method_sig_to_selector[method_signature] = method_selector
+        self.method_selector_to_sig[method_selector] = method_signature
+
+        if method_config.is_arc4_compliant():
```

Comment:
I'm still not happy with the `is_arc4_compliant` name. Taking a step back, is this method even necessary? It seems like you could get rid of this if statement without affecting anything -- if `is_arc4_compliant()` returns true, then you know that `approval_cond()` and `clear_state_cond()` will both return 1, so the code below this is equivalent anyway.

Reply:
The more I think, the more I felt absurd about that. I just deleted the `is_arc4_compliant` method, and related all constructions.

Do we also want to remove the `arc4_compliant`, or just rename it as `allow_all` or something alike?

Reply:
If there are no other uses for it, I think `arc4_compliant` can be removed

Reply:
just removed